sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,539 @@
+  return i * 2;
+}
+
+// Commands
+clang++ -fsycl lib.cpp -shared -o helpers.so
+clang++ -fsycl app.cpp -lhelpers -o a.out
+./a.out
+Output: 0 2 4 6…
+
+```
+It is possible to manually create `sycl::program` in both app and shared
+library, then use `link` SYCL API to get a single program and launch kernels
+using it. But it is not user-friendly and it is very different from regular
+C/C++ workflow.
+
+Another possible scenario - use functions defined in a pre-compiled device image
+provided by the user. Example:
+```
+// a.cpp
+SYCL_EXTERNAL void LibDeviceFunc();
+...
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for([]() { LibDeviceFunc(); });
+});
+
+// b.cpp
+/*no SYCL_EXTERNAL*/ void LibDeviceFunc() { ... }
+```
+We have a `SYCL_EXTERNAL` function `foo` called from a kernel, but the
+application defined only host version of this function. Then user adds device
+image with definition of `foo` to the fat object via special compiler option
+(like `-fsycl-add-targets`).
+
+The main purpose of this feature is to provide a mechanism which allows to
+link device code dynamically at runtime, such as in the scenarios above.
+
+## Requirements:
+User's device code that consists of some device API (`SYCL_EXTERNAL` functions),
+is compiled into some form and it is not linked statically with device code of
+application. It can be a shared library that contains some device code or a
+separate device image supplied with properties attached. This code is linked
+dynamically at run time with device code of a user's application in order to
+resolve dependencies.
+For this combination the following statements must be true:
+
+The presented dynamic device code linkage mechanism must:
+
+- Allow to represent the actual dynamically linked code as a device binary image
+  which can be:
+  - Embedded into a host shared object by standard SYCL compiler driver
+    invocation
+  - Embedded into a host binary or shared object using manual invocations of
+    SYCL tools such as `clang-offload-wrapper` and linker
+  - Loaded into memory via special API
+- Must not assume the actual format of the device code - e.g. that it is SPIR-V
+  or native device binary
+- Provide automatic runtime resolution of `SYCL_EXTERNAL` function references
+  within the SYCL app to their definitions (if found) within any suitable
+  dynamically linked device binary image
+- Support pointers to `SYCL_EXTERNAL` functions across the dynamic linkage
+  boundaries within the device code - taking a pointer, call through a pointer.
+- Specific code changes are not required, i.e. the mechanism of linking works
+  as close as possible to host shared libraries.
+
+## Design
+The overall idea:
+
+- Each device image is supplied with a list of imported and exported symbol
+  names through device image properties mechanism
+- Before JIT-compiling a device image DPC++ RT will check if device image has a
+  list of imported symbols and if it has, then RT will search for device images
```

Comment:
> RT will search for device images

where does it search?

Reply:
There is an internal storage inside `ProgramManager` runtime class. It stores all device images available during application runtime. I think the search will happen there.

Reply:
> It stores all device images available during application runtime

"available" where? how the list of images is populated?

Reply:
> "available" where?

Available to runtime library. This is basically all device images that were embedded into "fat" application binary and all loaded "fat" shared libraries.

> how the list of images is populated?

Each "fat" object is a host binary with device binary(s) embedded. We have a tool named `clang-offload-wrapper` that generates host "wrapper" object for device code that embedded into host object. This "wrapper" object contains a struct (descriptor) with information about wrapped code and device binaries themselves. The "wrapper" object also adds call to registration function which will be invoked when parent host binary is loaded. This "registration" function receives (descriptor) as a parameter. The runtime library defines registration function, so when each host object is loaded the registration function is called. AFAIK the implementation of this function in runtime library (I can grep it as `__sycl_register_lib`) adds device images represented by the descriptor parameter to internal storage in `ProgramManager` class. I think it is better explained in the doc https://github.com/intel/llvm/blob/sycl/sycl/doc/CompilerAndRuntimeDesign.md#fat-binary-creation-details . I'm also not aware of internals of `clang-offload-wrapper` tool.